ConcurrentModificationException applying 1.0.5 before publishing plugins

## What happened?

I'm applying revapi 1.0.5 right after applying `java-library` but before applying other plugins that configure publishing ([in this project](https://github.com/palantir/palantir-java-format/tree/22587955cff3e11687b8d9a11b19f72d06ae42be)):

```gradle
apply plugin: 'java-library'
apply plugin: 'com.palantir.revapi'
apply from: rootProject.file('gradle/publish-jar.gradle')
```

I then get this error:

<details>
<summary>
Stack trace
</summary>

<pre>
Caused by: org.gradle.api.internal.plugins.PluginApplicationException: Failed to apply plugin [id 'com.palantir.revapi']
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:167)
        at org.gradle.api.internal.plugins.DefaultPluginManager.apply(DefaultPluginManager.java:136)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.applyType(DefaultObjectConfigurationAction.java:129)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.access$200(DefaultObjectConfigurationAction.java:38)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction$3.run(DefaultObjectConfigurationAction.java:93)
        at org.gradle.api.internal.plugins.DefaultObjectConfigurationAction.execute(DefaultObjectConfigurationAction.java:152)
        at org.gradle.api.internal.project.AbstractPluginAware.apply(AbstractPluginAware.java:49)
        at org.gradle.api.internal.project.ProjectScript.apply(ProjectScript.java:41)
        at org.gradle.api.Script$apply$1.callCurrent(Unknown Source)
        at build_5noics6iamd59kl2acirukei.run(/Users/dsanduleac/code/external/palantir-java-format-2/palantir-java-format-api/build.gradle:2)
        at org.gradle.groovy.scripts.internal.DefaultScriptRunnerFactory$ScriptRunnerImpl.run(DefaultScriptRunnerFactory.java:91)
        ... 187 more
Caused by: java.util.ConcurrentModificationException
        at org.gradle.api.internal.collections.FilteredCollection$FilteringIterator.findNext(FilteredCollection.java:121)
        at org.gradle.api.internal.collections.FilteredCollection$FilteringIterator.next(FilteredCollection.java:140)
        at org.gradle.api.internal.DefaultDomainObjectCollection$IteratorImpl.next(DefaultDomainObjectCollection.java:479)
        at com.palantir.gradle.revapi.RevapiPlugin.apply(RevapiPlugin.java:60)
        at com.palantir.gradle.revapi.RevapiPlugin.apply(RevapiPlugin.java:34)
        at org.gradle.api.internal.plugins.ImperativeOnlyPluginTarget.applyImperative(ImperativeOnlyPluginTarget.java:43)
        at org.gradle.api.internal.plugins.RuleBasedPluginTarget.applyImperative(RuleBasedPluginTarget.java:51)
        at org.gradle.api.internal.plugins.DefaultPluginManager.addPlugin(DefaultPluginManager.java:181)
        at org.gradle.api.internal.plugins.DefaultPluginManager.access$300(DefaultPluginManager.java:51)
        at org.gradle.api.internal.plugins.DefaultPluginManager$AddPluginBuildOperation.run(DefaultPluginManager.java:276)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.plugins.DefaultPluginManager$2.execute(DefaultPluginManager.java:159)
        at org.gradle.api.internal.plugins.DefaultPluginManager$2.execute(DefaultPluginManager.java:156)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.apply(DefaultUserCodeApplicationContext.java:49)
        at org.gradle.api.internal.plugins.DefaultPluginManager.doApply(DefaultPluginManager.java:156)
        ... 197 more
</pre>

</details>


## What did you want to happen?

No crash.

--- a/changelog/@unreleased/pr-79.v2.yml
+++ b/changelog/@unreleased/pr-79.v2.yml
@@ -1,0 +1,6 @@
+type: fix
+fix:
+  description: Fixes `ConcurrentModificationException` thrown when `com.palantir.baseline`
+    (or more specifically `com.palantir.baseline-circleci`) is applied before `gradle-revapi`.
+  links:
+  - https://github.com/palantir/gradle-revapi/pull/79

--- a/src/main/java/com/palantir/gradle/revapi/RevapiPlugin.java
+++ b/src/main/java/com/palantir/gradle/revapi/RevapiPlugin.java
@@ -57,12 +57,12 @@ public final class RevapiPlugin implements Plugin<Project> {
             task.getOldGroupNameVersion().set(project.getProviders().provider(extension::oldGroupNameVersion));
         });
 
-        Jar jarTask = project.getTasks().withType(Jar.class).getByName(JavaPlugin.JAR_TASK_NAME);
-
-        project.getTasks().withType(RevapiJavaTask.class).forEach(task -> {
+        project.getTasks().withType(RevapiJavaTask.class).configureEach(task -> {
             task.dependsOn(allJarTasksIncludingDependencies(project, revapiNewApi));
             task.configManager().set(configManager);
             task.newApiDependencyJars().set(revapiNewApi);
+
+            Jar jarTask = project.getTasks().withType(Jar.class).getByName(JavaPlugin.JAR_TASK_NAME);
             task.newApiJars().set(jarTask.getOutputs().getFiles());
         });
 

--- a/src/main/java/com/palantir/gradle/revapi/RevapiPlugin.java
+++ b/src/main/java/com/palantir/gradle/revapi/RevapiPlugin.java
@@ -57,14 +57,12 @@ public final class RevapiPlugin implements Plugin<Project> {
             task.getOldGroupNameVersion().set(project.getProviders().provider(extension::oldGroupNameVersion));
         });
 
+        Jar jarTask = project.getTasks().withType(Jar.class).getByName(JavaPlugin.JAR_TASK_NAME);
+
         project.getTasks().withType(RevapiJavaTask.class).forEach(task -> {
             task.dependsOn(allJarTasksIncludingDependencies(project, revapiNewApi));
-
             task.configManager().set(configManager);
-
             task.newApiDependencyJars().set(revapiNewApi);
-
-            Jar jarTask = project.getTasks().withType(Jar.class).getByName(JavaPlugin.JAR_TASK_NAME);
             task.newApiJars().set(jarTask.getOutputs().getFiles());
         });
 

--- a/src/test/groovy/com/palantir/gradle/revapi/RevapiSpec.groovy
+++ b/src/test/groovy/com/palantir/gradle/revapi/RevapiSpec.groovy
@@ -395,6 +395,10 @@ class RevapiSpec extends IntegrationSpec {
 
     def 'does not throw exception when baseline-circleci is applied before this plugin'() {
         when:
+        addSubproject 'subproject',  """
+            apply plugin: '${TestConstants.PLUGIN_NAME}'
+        """
+
         buildFile << """
             buildscript {
                 repositories {
@@ -411,7 +415,6 @@ class RevapiSpec extends IntegrationSpec {
             // baseline-circleci is the bad plugin, but might as well test against all of baseline
             apply plugin: 'com.palantir.baseline'
             apply plugin: 'com.palantir.baseline-circleci'
-            apply plugin: '${TestConstants.PLUGIN_NAME}'
         """
 
         then:

--- a/src/test/groovy/com/palantir/gradle/revapi/RevapiSpec.groovy
+++ b/src/test/groovy/com/palantir/gradle/revapi/RevapiSpec.groovy
@@ -393,6 +393,31 @@ class RevapiSpec extends IntegrationSpec {
         assert stderr.contains('method void foo.Foo::setSomeProperty(java.lang.String)')
     }
 
+    def 'does not throw exception when baseline-circleci is applied before this plugin'() {
+        when:
+        buildFile << """
+            buildscript {
+                repositories {
+                    maven { url "https://palantir.bintray.com/releases" }
+                    jcenter()
+                    gradlePluginPortal()
+                }
+            
+                dependencies {
+                    classpath 'com.palantir.baseline:gradle-baseline-java:2.21.0'
+                }
+            }
+
+            // baseline-circleci is the bad plugin, but might as well test against all of baseline
+            apply plugin: 'com.palantir.baseline'
+            apply plugin: 'com.palantir.baseline-circleci'
+            apply plugin: '${TestConstants.PLUGIN_NAME}'
+        """
+
+        then:
+        runTasksSuccessfully("tasks")
+    }
+
     private String testMavenPublication() {
         return """
             publishing {

